packages/effect/src/Micro.ts

```diff
@@ -1653,23 +1653,48 @@ export const zip: {
   <A, E, R, A2, E2, R2>(
     self: Micro<A, E, R>,
     that: Micro<A2, E2, R2>,
-    options?:
-      | { readonly concurrent?: boolean | undefined }
-      | undefined
+    options?: { readonly concurrent?: boolean | undefined }
   ): Micro<[A, A2], E | E2, R | R2>
 } = dual((args) => isMicro(args[1]), <A, E, R, A2, E2, R2>(
   self: Micro<A, E, R>,
   that: Micro<A2, E2, R2>,
-  options?:
-    | { readonly concurrent?: boolean | undefined }
-    | undefined
+  options?: { readonly concurrent?: boolean | undefined }
 ): Micro<[A, A2], E | E2, R | R2> => {
-  if (options?.concurrent) {
-    return all([self, that], { concurrency: "unbounded" })
-  }
-  return flatMap(self, (a) => map(that, (a2) => [a, a2]))
+  return zipWith(self, that, (a, a2) => [a, a2], options)
 })
 
+/**
+ * The `Micro.zipWith` function combines two `Micro` effects and allows you to
+ * apply a function to the results of the combined effects, transforming them
+ * into a single value.
+ *
+ * @since 3.4.3
+ * @experimental
+ * @category zipping
+ */
+export const zipWith: {
+  <A2, E2, R2, A, B>(
+    that: Micro<A2, E2, R2>,
+    f: (a: A, b: A2) => B,
+    options?: { readonly concurrent?: boolean | undefined }
+  ): <E, R>(self: Micro<A, E, R>) => Micro<B, E2 | E, R2 | R>
+  <A, E, R, A2, E2, R2, B>(
+    self: Micro<A, E, R>,
+    that: Micro<A2, E2, R2>,
+    f: (a: A, b: A2) => B,
+    options?: { readonly concurrent?: boolean | undefined }
+  ): Micro<B, E2 | E, R2 | R>
+} = dual((args) => isMicro(args[1]), <A, E, R, A2, E2, R2, B>(
+  self: Micro<A, E, R>,
+  that: Micro<A2, E2, R2>,
+  f: (a: A, b: A2) => B,
+  options?: { readonly concurrent?: boolean | undefined }
+): Micro<B, E2 | E, R2 | R> =>
+  map(
+    all([self, that], { concurrency: options?.concurrent ? 2 : 1 }),
```

Comment:
You would only use `all` for the concurrent case. It adds a bit of overhead with all the concurrency book-keeping.